### DISCUSSION TITLE
Quarkus 3.16.x: Configuration mapping & validation not working

### DISCUSSION THREAD

**Discussion by joschi** [quarkusio/quarkus#44705] [Q&A] [answered]
Hey folks,

I have a fairly small Quarkus application with some custom configuration interfaces as described in [Mapping configuration to objects](https://quarkus.io/guides/config-mappings).

While the `@ConfigMapping` annotation and access the configuration properties via `@ConfigProperty` is working at runtime, none of the other annotations such as `@WithDefault` or any bean validation (`@NotNull` etc.) is working. `io.quarkus:quarkus-hibernate-validator` is on the compile and runtime class paths.

Example:
```java
@ConfigMapping(prefix = "jobs")
public interface JobsConfig {
	@NotNull
	@DurationMin(minutes = 1L)
	@WithDefault("PT5M")
	Duration runInterval();
}
```
Without `jobs.run-interval` in my configuration file, the application will fail with:
```text
Caused by: java.util.NoSuchElementException: Could not expand value jobs.run-interval in property ${jobs.run-interval}
```

In this case, the configuration property is being used in a scheduled job:
```java
@ApplicationScoped
public class MyScheduledJob {
	@Scheduled(every = "${jobs.run-interval}")
	Uni<Void> runScheduledJob() {
		// ...
	}
}
```

Any hints how to start debugging this?

- Quarkus 3.16.4
- Gradle 8.11.1
- Amazon Corretto 23

[a comment by an automated account is not shown]

**Comment by radcortez** [comment]
Because the mapping is not used anywhere (other than as a weak reference for an expanded default), Quarkus thinks it is not required and removed it from the final application. In this case, annotate the mapping with `@Unremovable`.

**Reply by joschi** [reply]
@radcortez Thanks for your answer!

> Quarkus thinks it is not required and removed it from the final application. In this case, annotate the mapping with `@Unremovable`.

Ooof, that's a pretty big caveat. Is this already documented anywhere and I missed it?

Adding `@Unremovable` seems to have resolved the issue with default values (`@WithDefault`) not being picked up but bean validation (`@Postitive`, `@MinDuration`) doesn't seem to work yet.

Is bean validation supposed to work when parsing configuration? https://quarkus.io/guides/config-mappings#validation indicates it should work.

`io.quarkus:quarkus-hibernate-validator` is on the compile and runtime class paths.

**Reply by gsmet** [reply]
@radcortez shouldn't we make the user-defined config mappings unremovable?

**Reply by gsmet** [reply]
@joschi could you assemble a small Maven reproducer? That usually helps.

**Reply by radcortez** [reply]
> @radcortez shouldn't we make the user-defined config mappings unremovable?

We did it that way because of this: https://github.com/quarkusio/quarkus/pull/19965.

If they are unremoveable, then you are forced to provide the configuration even if they are only injected in optional beans, but I do understand that it is inconvenient in this case.

One alternative solution is to provide the defaults separately so that the default is always available.

**Reply by joschi** [reply]
> could you assemble a small Maven reproducer? That usually helps.

Based on a fresh project created on code.quarkus.io with Gradle using the Kotlin DSL (since this is also what I am using):

[Download `config-validation-quarkus.zip`](https://github.com/user-attachments/files/17935968/config-validation-quarkus.zip)

```text
.
├── build.gradle.kts
├── gradle
│   └── wrapper
│       ├── gradle-wrapper.jar
│       └── gradle-wrapper.properties
├── gradle.properties
├── gradlew
├── gradlew.bat
├── settings.gradle.kts
└── src
    └── main
        ├── java
        │   └── org
        │       └── acme
        │           ├── ApplicationConfig.java
        │           └── AppScopedStartup.java
        └── resources
            └── application.properties
```

`ApplicationConfig.java`
```java
package org.acme;

import io.quarkus.arc.Unremovable;
import io.smallrye.config.ConfigMapping;
import jakarta.validation.constraints.Max;

@ConfigMapping(prefix = "application")
@Unremovable
public interface ApplicationConfig {
    @Max(10000)
    int port();
}
```

`AppScopedStartup.java`
```java
package org.acme;

import io.quarkus.runtime.Startup;
import jakarta.annotation.PostConstruct;
import jakarta.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.jboss.logging.Logger;

@Startup
@ApplicationScoped
public class AppScopedStartup {
    private static final Logger LOG = Logger.getLogger(AppScopedStartup.class);

    private final int port;

    public AppScopedStartup(@ConfigProperty(name = "application.port") int port) {
        this.port = port;
    }

    @PostConstruct
    void postConstruct() {
        // This shouldn't happen if the configuration was validated
        if (port > 10000) {
            throw new RuntimeException("NOOOOOOO! Value: " + port);
        }

        LOG.info("Value: " + port);
    }
}
```

`application.properties`
```properties
application.port=20000
```

Test run:
```text
❯ ./gradlew -q quarkusRun
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2024-11-27 15:44:43,680 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2024-11-27 15:44:43,752 ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:121)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
Caused by: java.lang.RuntimeException: NOOOOOOO! Value: 20000
        at org.acme.AppScopedStartup.postConstruct(AppScopedStartup.java:23)
        at org.acme.AppScopedStartup_Bean.doCreate(Unknown Source)
        at org.acme.AppScopedStartup_Bean.create(Unknown Source)
        at org.acme.AppScopedStartup_Bean.create(Unknown Source)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
        at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.c0(Unknown Source)
        at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
        at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:21)
        at org.acme.AppScopedStartup_ClientProxy.arc$delegate(Unknown Source)
        at org.acme.AppScopedStartup_ClientProxy.arc_contextualInstance(Unknown Source)
        at org.acme.AppScopedStartup_Observer_Synthetic_KSM1ACxw6RWzKMuh94trcYm-VKA.notify(Unknown Source)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:351)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:333)
        at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:80)
        at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:163)
        at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:114)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
        ... 11 more
```

**Reply by radcortez** [reply]
Yes, validation is only executed if the mapping is accessed (via injection, or programmatically). Maybe that is something we need to improve in the case where validation annotations are combined with `@Unremoveable`.

If I remember correctly, we did it that way because we didn't have the validator available when we started the config instance, but we made some improvements in that area @gsmet remember? Config has its own separate ready-to-use validator, so we can probably improve its behaviour.

**Reply by joschi** [reply]
@radcortez Can I somehow force the validation on my configuration beans as a workaround?

They're all required and not optional.

**Reply by radcortez** [reply, marked as answer]
Yes, do it like this:

```java
@Startup
@ApplicationScoped
public class AppScopedStartup {
    @Inject
    ApplicationConfig applicationConfig;

    @PostConstruct
    void postConstruct() {
        // This shouldn't happen if the configuration was validated
        if (applicationConfig.port() > 10000) {
            throw new RuntimeException("NOOOOOOO! Value: " + applicationConfig.port());
        }
    }
}
```

If you inject the mappings, the validation is triggered, and you don't even need the `@Unremoveable` annotation.

BTW, if you have a mapping, you might as well just use it instead of injecting `@ConfigProperty(name = "application.port") int port`

**Reply by joschi** [reply]
Awesome, that works. Thanks! 😄

**Reply by gsmet** [reply]
@radcortez should we add a note in the documentation? I suppose we have a section about config validation?

**Reply by radcortez** [reply]
A small section :)
https://quarkus.io/guides/config-mappings#validation

What is your idea? To note that you need `@Unremoveable` to force validation if the config class is not used? Another option is to always include mappings that have bean validation annotations.

**Reply by RomainWilbert** [reply]
how do i test this with Junit ? I was thinking if the configuration is missing, Quarkus should not start because of configuration  validation, but this does not work as expected

```
@QuarkusTest
@TestProfile(CremProcessingServiceConfKOTest.class)
public class CremProcessingServiceConfKOTest implements QuarkusTestProfile
{


   @Test
   void shouldFailBootstrapWithoutDestinataireConfig()
   {

      throw new AssertionError("Quarkus should have not started");
   }

   @Override
   public Map<String, String> getConfigOverrides()
   {
      return Map.of(
         "quarkus.config.locations", "src/test/resources/application-empty.properties"
      );
   }

}


   @Inject
   public CremProcessingService(MyConfigMapping conf)
   {
      this.conf = conf;
   }
```

**Reply by gsmet** [reply]
> Another option is to always include mappings that have bean validation annotations.

if it doesn’t cost us much, it’s actually a good idea.

**Reply by radcortez** [reply]
> how do i test this with Junit ? I was thinking if the configuration is missing, Quarkus should not start because of configuration validation, but this does not work as expected

Is the `CremProcessingService` being used somewhere? If not, Quarkus removes it, and all injections are also considered unused and removed. Can you please provide me with a full reproducer?

**Comment by RomainWilbert** [comment]
Hi @radcortez !

I guess there were several misunderstandings from my side : 

1. if getConfigOverrides() returns an empty map, this does not override existing configuration with empty values
2. if I override every mandatory configuration variable and set them to empty, the application crashes with an error as expected
3. When I run the Quarkus Test, the application does not start : so the Junit test fails with error, and does not execute any of the tests => I guess this is untestable, unless i start the app manually and check the process return code

So for now I gave up on testing that app crashes when mandatory configuration is not present.

**Reply by radcortez** [reply]
> 1. if getConfigOverrides() returns an empty map, this does not override existing configuration with empty values

No. Config is always overridden. 

> 2. if I override every mandatory configuration variable and set them to empty, the application crashes with an error as expected

Yes, this is the way to do it.

> 3. When I run the Quarkus Test, the application does not start : so the Junit test fails with error, and does not execute any of the tests => I guess this is untestable, unless i start the app manually and check the process return code

Yes, because the validation prevents Quarkus from starting. You can do it like we do here: https://github.com/quarkusio/quarkus/blob/e649d180ae7031805fff1d3f4a28c1b445a824b5/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/config/ConfigMappingInvalidTest.java#L42

These use internal APIs, which are subject to change without warning. Also, they are not intended to be used by applications directly.

I did try in an app, and it works as you would expect. The proper way to do it is to create an extension.
